### PR TITLE
Prepend the last modified project to MSBuildAllProjects

### LINKED ISSUE #1299: Make MSBuildAllProjects magically append on import
See https://github.com/dotnet/sdk/issues/277#issuecomment-257720464

Today, every file has to add itself manually and many don't.

@rainersigwald

### PR DESCRIPTION
**Pull request by jeffkl**
During evaluation, keep track of the last modified project in the import graph.
Prepend the path to the last modified project to `MSBuildAllProjects`
Stop setting `MSBuildAllProjects` in our targets
Add unit tests

Closes #1299

### CODE REVIEW

File: src/Build.UnitTests/Evaluation/Evaluator_Tests.cs [hunk 1]
```diff
@@ -1997,7 +1998,7 @@ public void AllEvaluatedPropertiesAndImports()
                 // the whole list.
                 // We have to dump it into a dictionary because AllEvaluatedProperties contains duplicates, but we're preparing to Properties,
                 // which doesn't, so we need to make sure that the final value in AllEvaluatedProperties is the one that matches.
-                foreach (ProjectProperty property in project.AllEvaluatedProperties.TakeWhile(property => property.Xml == null))
+                foreach (ProjectProperty property in project.AllEvaluatedProperties.Where(property => property.Xml == null))
```
Comment (rainersigwald) [inline src/Build.UnitTests/Evaluation/Evaluator_Tests.cs]: Why this change? If it's needed, please update the comment to remove the "don't need to scan the whole list" bit.

File: src/Build.UnitTests/Evaluation/Evaluator_Tests.cs [hunk 2]
```diff
@@ -4425,6 +4426,56 @@ public void ConditionWithShortCircuitAndErrorDoesNotFailBuild(string projectInne
...
+<Import Project=""{project1.ProjectFile}"" />
+<Import Project=""{project2.ProjectFile}"" />
+</Project>");
+
+                // Project1 and primary project last modified an hour ago, project2 is the newest
+                File.SetLastWriteTime(project1.ProjectFile, DateTime.Now.AddHours(-1));
+                File.SetLastWriteTime(project2.ProjectFile, DateTime.Now);
+                File.SetLastWriteTime(primaryProject.ProjectFile, DateTime.Now.AddHours(-1));
+
+
```
Comment (rainersigwald) [inline src/Build.UnitTests/Evaluation/Evaluator_Tests.cs]: nit: extra blank line

File: src/Build.UnitTests/Evaluation/Evaluator_Tests.cs [hunk 3]
```diff
@@ -4425,6 +4426,56 @@ public void ConditionWithShortCircuitAndErrorDoesNotFailBuild(string projectInne
...
+                Project project = new Project(primaryProject.ProjectFile, null, null);
+
+                project.GetPropertyValue(Constants.MSBuildAllProjectsPropertyName).ShouldStartWith(project2.ProjectFile);
```
Comment (rainersigwald) [inline src/Build.UnitTests/Evaluation/Evaluator_Tests.cs]: Validate no project1 path?

File: src/Build/Evaluation/Evaluator.cs [hunk 4]
```diff
@@ -2725,6 +2743,25 @@ private static string StringifyList(IList<string> strings)
...
+        {
+            if (_lastModifiedProject != null)
+            {
+                P oldValue = _data.GetProperty(Constants.MSBuildAllProjectsPropertyName);
+
+                P newValue = _data.SetProperty(
+                    Constants.MSBuildAllProjectsPropertyName,
+                    oldValue == null
+                        ? _lastModifiedProject.FullPath
+                        : String.Format(CultureInfo.CurrentCulture, "{0};{1}", _lastModifiedProject.FullPath, oldValue.EvaluatedValue), isGlobalProperty: false, mayBeReserved: false);
```
Comment (rainersigwald) [inline src/Build/Evaluation/Evaluator.cs]: Do we want `CurrentCulture`? Shouldn't it be invariant?

File: src/Build.UnitTests/Evaluation/Evaluator_Tests.cs (hunk 1 above)
Comment (jeffkl) [reply src/Build.UnitTests/Evaluation/Evaluator_Tests.cs]: This code assumes that all reserved properties are at the beginning of the list.  Originally I had added a new reserved property but it was at the end so this test broke.  Should I update the comment or revert this change?

File: src/Build/Evaluation/Evaluator.cs (hunk 4 above)
Comment (jeffkl) [reply src/Build/Evaluation/Evaluator.cs]: All other places in this file use `CurrentCulture`...

File: src/Build.UnitTests/Evaluation/Evaluator_Tests.cs (hunk 1 above)
Comment (rainersigwald) [reply src/Build.UnitTests/Evaluation/Evaluator_Tests.cs]: I'd just update the comment. It's test code; the perf benefit of an early stop doesn't strike me as worth much.

File: src/Build/Evaluation/Evaluator.cs (hunk 4 above)
Comment (AndyGerlicher) [reply src/Build/Evaluation/Evaluator.cs]: I'd change it. And just use string interpolation rather than `string.Format`.
